indexing-service/src/main/java/org/apache/druid/indexing/common/actions/RetrieveUsedSegmentsAction.java

```diff
@@ -113,11 +131,81 @@ public TypeReference<Collection<DataSegment>> getReturnTypeReference()
 
   @Override
   public Collection<DataSegment> perform(Task task, TaskActionToolbox toolbox)
+  {
+    // The DruidInputSource can be used to read from one datasource and write to another.
+    // In such a case, the race condition described in the class-level docs cannot occur,
+    // and the action can simply fetch all visible segments for the datasource and interval.
+    // Similarly, an MSQ replace could read from a different datasource.
```

Comment:
The race condition is always a possibility even when reading from a different datasource. There are really two possibilities here:
- There is no way to return a consistent set if this action is fired multiple times during the lifecycle of the task because the task does not have any lock on the other datasource.
- OR we do not care about the race condition because we read the segments just once.

Based on which of the above it really is, we should simplify the comment to something like:
```suggestion
    // When fetching segments for a datasource other than the one this task is writing to,
    // just return all visible segments as there is no way to ensure that the set of returned
    // segments is consistent throughout the lifecycle of this task
```

Reply:
Simplified, thanks